File: sdk/typescript/src/api/payments.ts
import { HTTPClient } from './http-client';
import { Invoice, PaymentMethod, PaginatedResponse } from '../types';

export class PaymentAPI {
  constructor(private client: HTTPClient) {}

  async createTransaction(data: {
    subscriberId: number;
    amount: number;
    currency?: string;
    gateway?: string;
    metadata?: Record<string, any>;
  }): Promise<any> {
    return this.client.request<any>({
      method: 'POST',
      endpoint: '/v1/payments/transactions',
      data,
    });
  }

  async getTransaction(transactionId: string): Promise<any> {
    return this.client.request<any>({
      method: 'GET',
      endpoint: `/v1/payments/transactions/${transactionId}`,
    });
  }

  async listTransactions(
    subscriberId?: number,
    status?: string,
    page: number = 1,
    pageSize: number = 50
  ): Promise<PaginatedResponse<any>> {
    const params: Record<string, string> = {
      page: page.toString(),
      page_size: pageSize.toString(),
    };

    if (subscriberId) {
      params.subscriber_id = subscriberId.toString();
    }
    if (status) {
      params.status = status;
    }

    return this.client.request<PaginatedResponse<any>>({
      method: 'GET',
      endpoint: '/v1/payments/transactions',
      params,
    });
  }

  async listInvoices(
    subscriberId?: number,
    status?: string,
    page: number = 1,
    pageSize: number = 50
  ): Promise<PaginatedResponse<Invoice>> {
    const params: Record<string, string> = {
      page: page.toString(),
      page_size: pageSize.toString(),
    };

    if (subscriberId) {
      params.subscriber_id = subscriberId.toString();
    }
    if (status) {
      params.status = status;
    }

    return this.client.request<PaginatedResponse<Invoice>>({
      method: 'GET',
      endpoint: '/v1/payments/invoices',
      params,
    });
  }

  async addPaymentMethod(data: {
    subscriberId: number;
    type: string;
    token: string;
    isDefault?: boolean;
  }): Promise<PaymentMethod> {
    return this.client.request<PaymentMethod>({
      method: 'POST',
      endpoint: '/v1/payments/methods',
      data,
    });
  }
}

File: sdk/rust/src/api/payments.rs
use crate::client::HTTPClient;
use crate::error::TelecomError;
use crate::types::{CreatePaymentRequest, PaymentTransaction, SubscriberList};
use std::collections::HashMap;

pub struct PaymentAPI {
    client: HTTPClient,
}

impl PaymentAPI {
    pub fn new(client: HTTPClient) -> Self {
        Self { client }
    }

    pub async fn create_transaction(&self, req: &CreatePaymentRequest) -> Result<PaymentTransaction, TelecomError> {
        self.client.post("/v1/payments/transactions", Some(req)).await
    }

    pub async fn get_transaction(&self, transaction_id: &str) -> Result<PaymentTransaction, TelecomError> {
        self.client
            .get(&format!("/v1/payments/transactions/{}", transaction_id), None)
            .await
    }

    pub async fn list_transactions(
        &self,
        subscriber_id: Option<i64>,
        status: Option<String>,
        page: i32,
        page_size: i32,
    ) -> Result<SubscriberList, TelecomError> {
        let mut params = HashMap::new();
        params.insert("page".to_string(), page.to_string());
        params.insert("page_size".to_string(), page_size.to_string());

        if let Some(subscriber_id) = subscriber_id {
            params.insert("subscriber_id".to_string(), subscriber_id.to_string());
        }
        if let Some(status) = status {
            params.insert("status".to_string(), status);
        }

        self.client.get("/v1/payments/transactions", Some(&params)).await
    }
}
